runtime/moonbase/src/lib.rs

```diff
@@ -1085,6 +1086,85 @@ impl pallet_moonbeam_orbiters::Config for Runtime {
 	type WeightInfo = pallet_moonbeam_orbiters::weights::SubstrateWeight<Runtime>;
 }
 
+/// Only callable after `set_validation_data` is called which forms this proof the same way
+fn relay_chain_state_proof() -> RelayChainStateProof {
+	let relay_storage_root = ParachainSystem::validation_data()
+		.expect("set in `set_validation_data`")
+		.relay_parent_storage_root;
+	let relay_chain_state =
+		ParachainSystem::relay_state_proof().expect("set in `set_validation_data`");
+	RelayChainStateProof::new(ParachainInfo::get(), relay_storage_root, relay_chain_state)
+		.expect("Invalid relay chain state proof, already constructed in `set_validation_data`")
+}
+
+pub struct BabeDataGetter;
+impl pallet_randomness::GetBabeData<BlockNumber, u64, Option<Hash>> for BabeDataGetter {
+	// Tolerate panic here because only ever called in inherent (so can be omitted)
+	fn get_relay_block_number() -> BlockNumber {
+		ParachainSystem::validation_data()
+			.expect("set in `set_validation_data`inherent => available before on_initialize")
+			.relay_parent_number
+	}
+	// Tolerate panic here because only ever called in inherent (so can be omitted)
+	fn get_relay_epoch_index() -> u64 {
+		relay_chain_state_proof()
+			.read_optional_entry(relay_chain::well_known_keys::EPOCH_INDEX)
+			.ok()
+			.flatten()
+			.unwrap_or_default()
+		// TODO: add back panic, only removing for testing now
+		//.expect("expected to be able to read epoch index from relay chain state proof")
+	}
```

Comment:
revert `unwrap_or_default()` before merge, just testing everything else works now

So if the EPOCH_INDEX isn't found in the relay proof, this would panic in the mandatory inherent. We said that's OK because we can exclude the inherent...but also `get_relay_epoch_index` is called in `Randomness::on_finalize` to verify the inherent is always included so if `get_relay_epoch_index` ever panics, no block would be valid.

I don't like returning an option from here because of this edge case, but it an alternative fix which would involve writing messy pallet logic which would just not update the RelayTime.relay_epoch_index if it cannot be read...